### ISSUE TITLE
The client fails to connect to the cluster if it doesn't know the hostname of a master [HZ-874]

### ISSUE THREAD

**Issue by slavagithub** [hazelcast/hazelcast#19301]
The client fails to connect to the cluster if it doesn't know the hostname of a master.
1. Start cluster with master node which knows its IP (for master: add to /etc/hosts line "172.49.1.1 master.host" with actual IP of the node and add to the configuration 
member-list:
          - master.host
 )  
2. Start the client with the address set to   172.49.1.1 (use actual IP of a master here)
**RESULT**
  Client fails to connect with "java.net.UnknownHostException: internal.not-known: nodename nor servname provided, or not known" error
**EXPECTED RESULT**
Client should be successfully connected.

Logs:
```
ug 10, 2021 1:26:40 PM com.hazelcast.client.impl.spi.ClientInvocationService
INFO: hz.client_1 [hostname-test] [5.0-SNAPSHOT] Running with 2 response threads, dynamic=true
Aug 10, 2021 1:26:41 PM com.hazelcast.core.LifecycleService
INFO: hz.client_1 [hostname-test] [5.0-SNAPSHOT] HazelcastClient 5.0-SNAPSHOT (20210810 - 57ce049, 70893c2) is STARTING
Aug 10, 2021 1:26:41 PM com.hazelcast.core.LifecycleService
INFO: hz.client_1 [hostname-test] [5.0-SNAPSHOT] HazelcastClient 5.0-SNAPSHOT (20210810 - 57ce049, 70893c2) is STARTED
WARNING: An illegal reflective access operation has occurred
WARNING: Illegal reflective access by com.hazelcast.internal.networking.nio.SelectorOptimizer (file:/Users/slava/.m2/repository/com/hazelcast/hazelcast-enterprise/5.0-SNAPSHOT/hazelcast-enterprise-5.0-20210810.075720-423.jar) to field sun.nio.ch.SelectorImpl.selectedKeys
WARNING: Please consider reporting this to the maintainers of com.hazelcast.internal.networking.nio.SelectorOptimizer
WARNING: Use --illegal-access=warn to enable warnings of further illegal reflective access operations
WARNING: All illegal access operations will be denied in a future release
Aug 10, 2021 1:26:41 PM com.hazelcast.client.impl.connection.ClientConnectionManager
INFO: hz.client_1 [hostname-test] [5.0-SNAPSHOT] Trying to connect to cluster: hostname-test
Aug 10, 2021 1:26:41 PM com.hazelcast.client.impl.connection.ClientConnectionManager
INFO: hz.client_1 [hostname-test] [5.0-SNAPSHOT] Trying to connect to [172.49.1.2]:5701
Aug 10, 2021 1:26:46 PM com.hazelcast.client.impl.connection.ClientConnectionManager
WARNING: hz.client_1 [hostname-test] [5.0-SNAPSHOT] Exception during initial connection to [172.49.1.2]:5701: com.hazelcast.core.HazelcastException: java.io.IOException: null to address /172.49.1.2:5701
Aug 10, 2021 1:26:46 PM com.hazelcast.client.impl.connection.ClientConnectionManager
INFO: hz.client_1 [hostname-test] [5.0-SNAPSHOT] Trying to connect to [172.49.1.1]:5701
Aug 10, 2021 1:26:51 PM com.hazelcast.client.impl.connection.ClientConnectionManager
WARNING: hz.client_1 [hostname-test] [5.0-SNAPSHOT] Exception during initial connection to [172.49.1.1]:5701: com.hazelcast.core.HazelcastException: java.io.IOException: null to address /172.49.1.1:5701
Aug 10, 2021 1:26:51 PM com.hazelcast.client.impl.connection.ClientConnectionManager
INFO: hz.client_1 [hostname-test] [5.0-SNAPSHOT] Trying to connect to [127.0.0.1]:5701
Aug 10, 2021 1:26:51 PM com.hazelcast.client.impl.connection.tcp.TcpClientConnection
WARNING: hz.client_1 [hostname-test] [5.0-SNAPSHOT] ClientConnection{alive=false, connectionId=1, channel=NioChannel{/127.0.0.1:50701->/127.0.0.1:5701}, remoteAddress=null, lastReadTime=2021-08-10 13:26:51.117, lastWriteTime=2021-08-10 13:26:51.100, closedTime=2021-08-10 13:26:51.123, connected server version=null} closed. Reason: Failed to authenticate connection
com.hazelcast.core.HazelcastException: java.net.UnknownHostException: internal.not-known: nodename nor servname provided, or not known
	at com.hazelcast.client.impl.protocol.codec.builtin.CustomTypeFactory.createAddress(CustomTypeFactory.java:65)
	at com.hazelcast.client.impl.protocol.codec.custom.AddressCodec.decode(AddressCodec.java:58)
	at com.hazelcast.client.impl.protocol.codec.builtin.CodecUtil.decodeNullable(CodecUtil.java:55)
	at com.hazelcast.client.impl.protocol.codec.ClientAuthenticationCodec.decodeResponse(ClientAuthenticationCodec.java:218)
	at com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager.authenticateOnCluster(TcpClientConnectionManager.java:881)
	at com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager.getOrConnectToAddress(TcpClientConnectionManager.java:612)
	at com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager.lambda$doConnectToCandidateCluster$3(TcpClientConnectionManager.java:508)
	at com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager.connect(TcpClientConnectionManager.java:462)
	at com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager.doConnectToCandidateCluster(TcpClientConnectionManager.java:508)
	at com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager.doConnectToCluster(TcpClientConnectionManager.java:414)
	at com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager.connectToCluster(TcpClientConnectionManager.java:375)
	at com.hazelcast.client.impl.connection.tcp.TcpClientConnectionManager.start(TcpClientConnectionManager.java:324)
	at com.hazelcast.client.impl.clientside.HazelcastClientInstanceImpl.start(HazelcastClientInstanceImpl.java:373)
	at com.hazelcast.client.HazelcastClient.constructHazelcastClient(HazelcastClient.java:460)
	at com.hazelcast.client.HazelcastClient.newHazelcastClientInternal(HazelcastClient.java:416)
	at com.hazelcast.client.HazelcastClient.newHazelcastClient(HazelcastClient.java:136)
	at com.hazelcast.qe.hostnames.TestUtil.getClient(TestUtil.java:95)
	at com.hazelcast.qe.hostnames.DockerTest.clusterIsOkIfConfiguredAs(DockerTest.java:53)
	at com.hazelcast.qe.issues.Issue8.testClientCanConnectToClusterWhenHostnameIsNotKnown(Issue8.java:10)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
```

**Comment by sancar** [comment]
Ok. We need to be careful with this one. Note that fixing the authentication is not enough. After authentication response, the member will send the member list containing the unknown host-names. 
And the client whether it is smart or not, will use those host-names to connect to cluster when needed. In this case, it will fail of course. 
So it looks to me add this setup can not be supported with the promises we give to the customer. Since it can not work properly having exception at the first start looks more preferable.

**Comment by slavagithub** [comment]
Just to keep information in one place: Slack discussion: 

Josef Cacek Jan 27th at 10:02 AM
To me the full address resolution only makes sense in smart clients. If I'm a dummy one I don't care what's the cluster topology at all, what I want to do is:
connect to the first available address where anyone is listening;
provide the listening member my credentials;
use the member to get/put the data
heart_eyes
smile
relieved




5 replies

Sancar Koyunlu  5 days ago
One thing is not aligned with this description is that the dummy clients should be able to connect another member when the first connected member is shutdown.
I am not saying that this is a problem for removing resoltuion on auth response yet. But we provide this feature on our dummy clients, which forces them to be aware of the cluster topology.

Sancar Koyunlu  5 days ago
May be we need even dummier client config for this one.

Josef Cacek  5 days ago
Sure, I mean the fallback and retry mechanisms should still work, but only with addresses configured on the client.

Sancar Koyunlu  5 days ago
Not sure if documented this way, but the current implementaion of dummy client listens the cluster and uses an address from config+latest member list to find a next candidate to connect when needed.
Also there API's like get all members on cluster service and executeOnAllMembers . Methods like executeOnAllMembers works with uuids instead of addresses for some time. So address might not be needed in most of the places. But dummy has to know the member-list(for uuid's, and sometimes to return to the user from API ).

Sancar Koyunlu  5 days ago
So in short, we can improve the client. Not impossible, but tricky.
